src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py

```diff
@@ -0,0 +1,418 @@
+            }
+
+            try:
+                clone_source["group"] = self.metadata_mgr.get_option("source", "group")
+            except MetadataMgrException as me:
+                if me.errno == -errno.ENOENT:
+                    pass
+                else:
+                    raise
+        except MetadataMgrException as me:
+            raise VolumeException(-errno.EINVAL, "error fetching subvolume metadata")
+        return clone_source
+
+    @property
+    def status(self):
+        state = SubvolumeStates.from_value(self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_STATE))
+        subvolume_type = self.type
+        subvolume_status = {
+            'state' : state.value
+        }
+        if not SubvolumeOpSm.is_complete_state(state) and subvolume_type == SubvolumeTypes.TYPE_CLONE:
+            subvolume_status["source"] = self._get_clone_source()
+        return subvolume_status
+
+    @property
+    def state(self):
+        return SubvolumeStates.from_value(self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_STATE))
+
+    @state.setter
+    def state(self, val):
+        state = val[0].value
+        flush = val[1]
+        self.metadata_mgr.update_global_section(MetadataManager.GLOBAL_META_KEY_STATE, state)
+        if flush:
+            self.metadata_mgr.flush()
+
+    @property
+    def type(self):
+        return SubvolumeTypes.from_value(self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_TYPE))
+
+    def remove(self, retainsnaps=False):
+        if self.list_snapshots():
+            if not retainsnaps:
+                raise VolumeException(-errno.ENOTEMPTY, "subvolume '{0}' has snapshots".format(self.subvolname))
+            self.trash_incarnation_dir()
+            self.metadata_mgr.update_global_section(MetadataManager.GLOBAL_META_KEY_PATH, "")
+            self.metadata_mgr.update_global_section(MetadataManager.GLOBAL_META_KEY_STATE, SubvolumeStates.STATE_RETAINED.value)
+            self.metadata_mgr.flush()
+        else:
+            self.trash_base_dir()
+
+    def resize(self, newsize, noshrink):
+        subvol_path = self.path
+        return self._resize(subvol_path, newsize, noshrink)
+
+    def info(self):
+        if self.state != SubvolumeStates.STATE_RETAINED:
+            return self._info()
+
+        etype = self.type
+        features = self.features
+
+        return {'type': etype.value, 'features': features, 'state': SubvolumeStates.STATE_RETAINED.value}
+
+    def snapshot_path(self, snapname):
+        return os.path.join(self.base_path,
+                            self.vol_spec.snapshot_dir_prefix.encode('utf-8'),
+                            snapname.encode('utf-8'))
+
+    @staticmethod
+    def is_valid_uuid(uuid_str):
+        try:
+            uuid.UUID(uuid_str)
+            return True
+        except ValueError:
+            return False
+
+    def snapshot_data_path(self, snapname):
```

Comment:
Design-wise accessing snapshots does not require a valid subvolume object. Should snapshots be treated as a separate entity altogether?

Reply:
Snapshots in V1 depend on the path component from the meta file, in V2 are not dependent on the meta file for lookups.

Snapshots are dependent on updating the meta file for attach/detach operations as such.

As the meta file is managed via the subvolume, and V2 is derived from V1, I would leave snapshots as is for now.